### PR TITLE
Added SDIRK Solvers

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

Solves part of #2177

### CODE REVIEW

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas I added solvers for SDIRK methods. Can you please review this ??

File: src/algorithms.jl [hunk 1]
```diff
@@ -2293,8 +1430,8 @@ const MultistepAlgorithms = Union{
     ABDF2,
     AB3, AB4, AB5, ABM32, ABM43, ABM54}
 
-const SplitAlgorithms = Union{CNAB2, CNLF2, SBDF,
-    KenCarp3, KenCarp4, KenCarp47, KenCarp5, KenCarp58, CFNLIRK3}
+#KenCarp3, KenCarp4, KenCarp47, KenCarp5, KenCarp58, CFNLIRK3
+const SplitAlgorithms = Union{CNAB2, CNLF2, SBDF}
```
Comment (ChrisRackauckas) [inline src/algorithms.jl:1434]: Everything using this should instead use `issplit(alg)`

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas How do I solve this ImplicitEuler Undefined error ?? I have defined the ImplicitEuler inside algorithms.jl in OrdinaryDiffEqSDIRK/src but it still throws this error

File: src/OrdinaryDiffEq.jl [hunk 2]
```diff
@@ -149,14 +149,13 @@ include("generic_rosenbrock.jl")
...
 include("caches/high_order_rk_caches.jl")
-include("caches/sdirk_caches.jl")
 include("caches/firk_caches.jl")
-include("caches/kencarp_kvaerno_caches.jl")
 include("caches/linear_caches.jl")
 include("caches/linear_nonlinear_caches.jl")
 include("caches/rosenbrock_caches.jl")
 include("caches/adams_bashforth_moulton_caches.jl")
 include("caches/nordsieck_caches.jl")
+include("../lib/OrdinaryDiffEqSDIRK/src/sdirk_caches.jl")
```
Comment (ChrisRackauckas) [inline src/OrdinaryDiffEq.jl:158]:
```suggestion
```

File: lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl [hunk 3]
```diff
@@ -0,0 +1,30 @@
+module OrdinaryDiffEqSDIRK
+
+import OrdinaryDiffEq: alg_order, calculate_residuals!,
+                       initialize!, perform_step!, @unpack, unwrap_alg,
+                       calculate_residuals, alg_extrapolates,
+                       OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqAdaptivePartitionedAlgorithm,
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl:8]:
```suggestion
                       OrdinaryDiffEqNewtonAdaptiveAlgorithm,
```

File: src/OrdinaryDiffEq.jl [hunk 4]
```diff
@@ -275,6 +267,20 @@ include("../lib/OrdinaryDiffEqVerner/src/OrdinaryDiffEqVerner.jl")
 using ..OrdinaryDiffEqVerner
 export Vern6, Vern7, Vern8, Vern9
 
+include("../lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl")
+using ..OrdinaryDiffEqSDIRK
```
Comment (ChrisRackauckas) [inline src/OrdinaryDiffEq.jl:271]:
```suggestion
using ..OrdinaryDiffEqSDIRK
import .OrdinaryDiffEqSDIRK: ImplicitEulerConstantCache
```

File: lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                       OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqNewtonAlgorithm,
+                       DEFAULT_PRECS, NLNewton,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new, _ode_interpolant,
+                       markfirststage!, trivial_limiter!, _ode_interpolant!, 
+                       _ode_addsteps!, build_nlsolver, nlsolve!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl:15]:
```suggestion
                       _ode_addsteps!, build_nlsolver, nlsolve!, nlsolvefail
```

File: lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                       OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqNewtonAlgorithm,
+                       DEFAULT_PRECS, NLNewton,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new, _ode_interpolant,
+                       markfirststage!, trivial_limiter!, _ode_interpolant!, 
+                       _ode_addsteps!, build_nlsolver, nlsolve!, nlsolvefail
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl:15]:
```suggestion
                       _ode_addsteps!, build_nlsolver, nlsolve!, nlsolvefail, isnewton
```

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas What is I exactly?? I get a `I` not defined error here

**Comment by ChrisRackauckas** [comment]
Identity matrix, using LinearAlgebra: I

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas How can I deal with the error_constant UnDef error?? It's related to the nlsolve/newton.jl and bdf_utils.jl.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas There are some precision errors left here and an UnDef error saying IMEXEuler is not defined

File: lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+include("kencarp_kvaerno_caches.jl")
+include("sdirk_perform_step.jl")
+include("kencarp_kvaerno_perform_step.jl")
+include("sdirk_tableaus.jl")
+
+export ImplicitEuler, ImplicitMidpoint, Trapezoid, TRBDF2, SDIRK2, SDIRK22,
+       Kvaerno3, KenCarp3, Cash4, Hairer4, Hairer42, SSPSDIRK2, Kvaerno4,
+       Kvaerno5, KenCarp4, KenCarp47, KenCarp5, KenCarp58, ESDIRK54I8L2SA, SFSDIRK4,
+       SFSDIRK5, CFNLIRK3, SFSDIRK6, SFSDIRK7, SFSDIRK8, Kvaerno5, KenCarp4, KenCarp5,
+       SFSDIRK4, SFSDIRK5, CFNLIRK3, SFSDIRK6,
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl:34]:
```suggestion
       SFSDIRK4, SFSDIRK5, CFNLIRK3, SFSDIRK6, IMEXEuler, IMEXEulerARK,
```

File: lib/OrdinaryDiffEqBDF/src/OrdinaryDiffEqBDF.jl [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+                       NLNewton, alg_cache, _vec, _reshape, @cache, 
+                       isfsal, full_cache, build_nlsolver,
+                       nlsolve!, nlsolvefail, isnewton,
+                       constvalue, _unwrap_val, 
+                       DIRK, set_new_W!,
+                       du_alias_or_new, trivial_limiter!, 
+                       ImplicitEulerConstantCache, 
+                       ImplicitEulerCache, COEFFICIENT_MULTISTEP,
+                       markfirststage!, UJacobianWrapper, mul!
+using TruncatedStacktraces, MuladdMacro, MacroTools, FastBroadcast, RecursiveArrayTools
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqBDF/src/OrdinaryDiffEqBDF.jl:21]:
```suggestion
                       markfirststage!, UJacobianWrapper, mul!,
                       issplit, qsteady_min_default, qsteady_max_default,
                       get_current_alg_order, get_current_adaptive_order
using TruncatedStacktraces, MuladdMacro, MacroTools, FastBroadcast, RecursiveArrayTools
```

**Comment by ChrisRackauckas** [comment]
DImplicitEuler, DABDF2, and DFBDF should probably go with the other BDF methods. But that could probably be a next PR in order to let this one finish.

File: lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+                       OrdinaryDiffEqNewtonAdaptiveAlgorithm,
+                       OrdinaryDiffEqNewtonAlgorithm,
+                       DEFAULT_PRECS, NLNewton,
+                       OrdinaryDiffEqAdaptiveAlgorithm, CompiledFloats, uses_uprev,
+                       alg_cache, _vec, _reshape, @cache, isfsal, full_cache,
+                       constvalue, _unwrap_val, du_alias_or_new, _ode_interpolant,
+                       markfirststage!, trivial_limiter!, _ode_interpolant!, 
+                       UJacobianWrapper, set_new_W!, dolinsolve, get_W,
+                       build_nlsolver, nlsolve!, nlsolvefail, isnewton,
+                       COEFFICIENT_MULTISTEP, mul!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSDIRK/src/OrdinaryDiffEqSDIRK.jl:17]:
```suggestion
                       COEFFICIENT_MULTISTEP, mul!, isesdirk, issplit,
                       ssp_coefficient
```

File: lib/OrdinaryDiffEqBDF/src/OrdinaryDiffEqBDF.jl [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                       isfsal, full_cache, build_nlsolver,
+                       nlsolve!, nlsolvefail, isnewton,
+                       constvalue, _unwrap_val, 
+                       DIRK, set_new_W!,
+                       du_alias_or_new, trivial_limiter!, 
+                       ImplicitEulerConstantCache, 
+                       ImplicitEulerCache, COEFFICIENT_MULTISTEP,
+                       markfirststage!, UJacobianWrapper, mul!,
+                       issplit, qsteady_min_default, qsteady_max_default,
+                       get_current_alg_order, get_current_adaptive_order
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqBDF/src/OrdinaryDiffEqBDF.jl:22]:
```suggestion
                       get_current_alg_order, get_current_adaptive_order,
                       default_controller, stepsize_controller!, step_accept_controller!, 
                       step_reject_controller!, post_newton_controller!
```

**Pull request merged**